MapboxNavigationTests/NavigationMapViewTests.swift

```diff
@@ -103,6 +104,57 @@ class NavigationMapViewTests: XCTestCase, MGLMapViewDelegate {
         mapView!.removeAnnotations(mapView!.annotations ?? [])
         XCTAssertNil(mapView!.annotations)
     }
+    
+    func setUpVanishingRouteLine() -> Route {
+        let routeData = Fixture.JSONFromFileNamed(name: "vanish_point_test")
+        let routeOptions = NavigationRouteOptions(coordinates: [
+            CLLocationCoordinate2DMake(-122.5237734,37.9753973),
+            CLLocationCoordinate2DMake(-122.5264995,37.9709171)
+        ])
+        let decoder = JSONDecoder()
+        decoder.userInfo[.options] = routeOptions
+        var testRoute: Route?
+        XCTAssertNoThrow(testRoute = try decoder.decode(Route.self, from: routeData))
+        guard let route = testRoute else {
+            preconditionFailure("Route is invalid.")
+        }
+        return route
+    }
+    
+    func testUpdateVanishingPoint() {
+        let route = setUpVanishingRouteLine()
+        let navigationMapView = NavigationMapView(frame: CGRect(origin: .zero, size: .iPhone6Plus), styleURL: Fixture.blankStyle)
+        let shape = route.shape!
+        let targetPoint = Turf.mid(shape.coordinates[6], shape.coordinates[7])
+        //which is between route.legs[0].steps[1].shape!.coordinates[3] and  route.legs[0].steps[1].shape!.coordinates[4]
+        let traveledCoordinates = Array(route.legs[0].steps[1].shape!.coordinates[0...3])
+        let stepTraveledDistanceSep = navigationMapView.calculateGranularDistances(traveledCoordinates).distance
+
+        let testRouteProgress: RouteProgress = RouteProgress(route: route, routeIndex: 0, options: routeOptions, legIndex: 0, spokenInstructionIndex: 0)
+        testRouteProgress.currentLegProgress = RouteLegProgress(leg: route.legs[0], stepIndex: 1, spokenInstructionIndex: 0)
+        testRouteProgress.currentLegProgress.currentStepProgress = RouteStepProgress(step: route.legs[0].steps[1], spokenInstructionIndex: 0)
+        testRouteProgress.currentLegProgress.currentStepProgress.distanceTraveled = stepTraveledDistanceSep
+
+        navigationMapView.initPrimaryRoutePoints(route: route)
+        navigationMapView.updateUpcomingRoutePointIndex(routeProgress: testRouteProgress)
+        navigationMapView.updateTraveledRouteLine(targetPoint)
+
+        //TODO: the expected fraction should be provided
+        let expectedTraveledFraction = 0.06383308537010246
+        XCTAssertEqual(navigationMapView.fractionTraveled, expectedTraveledFraction)
+    }
+    
+    func testParseRoutePoints() {
+        let route = setUpVanishingRouteLine()
+        let navigationMapView = NavigationMapView(frame: CGRect(origin: .zero, size: .iPhone6Plus), styleURL: Fixture.blankStyle)
+        
+        navigationMapView.initPrimaryRoutePoints(route: route)
+        let nestedList = navigationMapView.primaryRoutePoints!.nestedList
+        let flatList = navigationMapView.primaryRoutePoints!.flatList
+        XCTAssertEqual(nestedList.first!.count, 8)
+        XCTAssertEqual(flatList.count, 33)
```

Comment:
Please try not to use force-unwrapping whenever possible. In case of running tests locally they will fail instead of crashing if there are any errors. E.g.:
```swift
let nestedList = navigationMapView.primaryRoutePoints?.nestedList
let flatList = navigationMapView.primaryRoutePoints?.flatList
XCTAssertEqual(nestedList?.first?.count, 8)
XCTAssertEqual(flatList?.count, 33)
```

Please also add verification check to make sure that `navigationMapView.primaryRouteLineGranularDistances` contains data you really expect.

Reply:
Thanks for the suggestion, I'll work on it.